File: app/src/main/java/xyz/rodit/snapmod/features/opera/OperaModelModifier.java
package xyz.rodit.snapmod.features.opera;

import java.util.ArrayList;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;
import xyz.rodit.snapmod.features.Feature;
import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.MapKey;
import xyz.rodit.snapmod.mappings.ParamsMap;

public class OperaModelModifier extends Feature {

    private final List<OperaPlugin> plugins = new ArrayList<>();

    public OperaModelModifier(FeatureContext context) {
        super(context);
    }

    @Override
    protected void init() {
        plugins.add(new MenuModifier(context));
        plugins.add(new SnapDurationModifier(context));
    }

    @Override
    protected void performHooks() {
        // Modify opera model map on insert.
        ParamsMap.put.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                String key = MapKey.wrap(param.args[0]).getName();
                plugins.stream().filter(p -> p.isEnabled() && p.shouldOverride(key)).forEach(
                        p -> param.args[1] = p.override(key, param.args[1])
                );
            }
        });
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/opera/OperaModelModifier.kt
package xyz.rodit.snapmod.features.opera

import xyz.rodit.snapmod.features.Feature
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.MapKey
import xyz.rodit.snapmod.mappings.ParamsMap
import xyz.rodit.snapmod.util.before

class OperaModelModifier(context: FeatureContext) : Feature(context) {

    private val plugins: MutableList<OperaPlugin> = ArrayList()

    override fun init() {
        plugins.add(MenuModifier(context))
        plugins.add(SnapDurationModifier(context))
    }

    override fun performHooks() {
        // Modify opera model map on insert.
        ParamsMap.put.before {
            val params = ParamsMap.wrap(it.thisObject)
            val key = MapKey.wrap(it.args[0]).name
            plugins.filter { p -> p.isEnabled && p.shouldOverride(params, key) }
                .forEach { p -> it.args[1] = p.override(params, key, it.args[1]) }
        }
    }
}